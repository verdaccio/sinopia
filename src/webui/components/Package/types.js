/**
 * @prettier
 * @flow
 */

export interface IProps {
  name: string;
  version: string;
  time: string;
  author: IAuthor;
  description?: string;
  keywords?: string[];
  license?: string;
  homepage: string;
  bugs: IBugs;
  dist: IDist;
}

export interface IAuthor {
  name: string;
  avatar: string;
  email: string;
}

export interface IBugs {
  url: string;
}
export interface IDist {
  unpackedSize: number;
}
