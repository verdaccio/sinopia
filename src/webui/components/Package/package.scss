@import '../../styles/variables';
@import '../../styles/mixins';

.package {
  .header {
    display: flex;
    align-items: center;
    margin: 10px 0 0;
  }

  .footer {
    display: flex;
    p.description {
      width: 100%;
      margin-bottom: 0;
      font-size: $font-size-sm;
      color: $grey-dark;
    }
  }

  .details {
    display: flex;
    font-size: 80%;
    color: $grey-light;
    padding-top: 5px;
    .license {
      width: 20%;
      text-align: right;
    }

    .homepage {
      width: 80%;
    }
  }

  > a {
    display: block;
    position: relative;
    color: inherit;
    margin: 0;
    padding: 10px 0;
    cursor: pointer;
    text-decoration: none;

    .title {
      width: 100%;

      h1 {
        font-size: $font-size-md;
        margin: 0;

        :global {
          .el-tag {
            margin-left: 5px;
          }
        }
      }
    }

    .author {
      color: $grey-light;
      font-size: inherit;
      width: 30%;
      text-align: right;
    }

    &:hover {
      &::before {
        content: '';
        display: block;
        background: rgba(255, 255, 255, 0.7);
        @include fullSize;
      }

      &::after {
        display: block;
        position: absolute;
        top: 50%;
        right: 0;
        left: 0;
        text-align: center;
        transform: translateY(-50%);
        font-size: $font-size-md;
      }
    }
  }
}
