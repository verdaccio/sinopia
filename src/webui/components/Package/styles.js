/**
 * @prettier
 * @flow
 */

import styled, { css } from 'react-emotion';
import { Link } from 'react-router-dom';
import { default as Photo } from '@material-ui/core/Avatar';
import { default as Ico } from '../Icon';

import mq from '../../utils/styles/media';
import { ellipsis } from '../../utils/styles/mixings';
import colors from '../../utils/styles/colors';

import Label from '../Label';

// HEADER
export const Header = styled.div`
  && {
    display: flex;
    flex-direction: column;
    padding: 0 0 5px 0;
  }
`;

export const Name = styled.span`
  && {
    ${ellipsis('50%')};
  }
`;

export const MainInfo = styled.span`
  && {
    font-size: 16px;
    font-weight: 600;
    line-height: 30px;
    flex: 1;
    color: #3a8bff;
    padding: 0 10px 0 0;
    overflow: hidden;
    display: flex;
    align-items: center;
    :hover {
      ${Name} {
        text-decoration: underline;
      }
    }
  }
`;

export const OverviewItem = styled.span`
  && {
    display: flex;
    align-items: center;
    margin: 0 0 5px 0;
    color: ${colors.greyLight};
  }
`;

export const Overview = styled.span`
  && {
    position: relative;
    display: flex;
    flex-direction: column;
  }
`;

export const Version = styled.span`
  && {
    font-size: 12px;
    padding: 0 0 0 10px;
    margin: 0 0 0 5px;
    color: #9f9f9f;
    position: relative;
    ${ellipsis('100%')};
    :before {
      content: 'â€¢';
      position: absolute;
      left: 0;
    }
  }
`;

export const Icon = styled(Ico)`
  && {
    margin: 1px 5px 0 0;
    fill: ${colors.greyLight};
  }
`;

export const Published = styled.span`
  && {
    display: none;
    color: ${colors.greyLight};
    ${({ modifiers }) => modifiers};
  }
`;

// Content
export const Field = styled.div`
  && {
    padding: 0 0 5px 0;
  }
`;

export const Content = styled.div`
  && {
    ${Field} {
      :last-child {
        padding: 0;
      }
    }
  }
`;

export const Text = styled(Label)`
  && {
    color: #908ba1;
  }
`;

export const Details = styled.span`
  && {
    margin-left: 5px;
    line-height: 14px;
    display: flex;
    flex-direction: column;
  }
`;

export const Author = styled.div`
  && {
    display: flex;
    align-items: center;
  }
`;

export const Avatar = styled(Photo)`
  && {
    width: 30px;
    height: 30px;
    background: #4b5e40;
    font-size: 15px;
  }
`;

export const Description = styled.div`
  && {
    margin: 5px 0;
    max-height: 100px;
    text-overflow: ellipsis;
  }
`;

// Footer
export const Footer = styled.div`
  && {
    display: none;
    padding: 5px 0 0 0;
  }
`;

// Container
export const Wrapper = styled(Link)`
  && {
    font-size: 12px;
    background-color: #fff;
    margin: 0 0 15px 0;
    transition: box-shadow 0.15s;
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.15);
    border-radius: 3px;
    padding: 10px;
    text-decoration: none;
    display: block;
    color: #2f273c;
    ${mq.medium(css`
      ${Header} {
        flex-direction: row;
        justify-content: space-between;
        align-items: center;
      }
      ${OverviewItem} {
        margin: 0 0 0 10px;
      }
      ${Overview} {
        flex-direction: row;
        ${OverviewItem} {
          :first-child {
            margin: 0;
          }
        }
      }
      ${Footer} {
        display: block;
      }
      ${Published} {
        display: inline-block;
      }
    `)};
  }
`;
