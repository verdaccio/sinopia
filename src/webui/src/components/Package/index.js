import React from 'react';
import PropTypes from 'prop-types';
import {Tag} from 'element-react';
import {Link} from 'react-router-dom';
import isNil from 'lodash/isNil';
import {formatDateDistance} from '../../utils/DateUtils';

import classes from './package.scss';

export default class Package extends React.Component {
  static propTypes = {
    package: PropTypes.object
  }

  render() {
    const {package: pkg} = this.props;

    return (
      <section className={classes.package}>
        <Link to={`detail/${pkg.name}`}>
          <div className={classes.header}>
            {this.renderTitle(pkg)}
            {this.renderAuthor(pkg)}
          </div>
          <div className={classes.footer}>
            {this.renderDescription(pkg)}
          </div>
          <div className={classes.details}>
            {this.renderPublished(pkg)}
            {this.renderLicense(pkg)}
          </div>
        </Link>
      </section>
    );
  }

  renderPublished(pkg) {
      return (<div className={classes.homepage}>
        {pkg.time ? `Published ${formatDateDistance(pkg.time)} ago`: ''}
      </div>);
  }

  renderLicense(pkg) {
    if (pkg.license) {
      return (<div className={classes.license}>
        {pkg.license}
      </div>);
    }

    return null;
  }

  renderDescription(pkg) {
    return (
      <p className={classes.description}>
        {pkg.description}
      </p>
    );
  }

  renderTitle(pkg) {
    return (
      <div className={classes.title}>
        <h1>
          {pkg.name} {this.renderTag(pkg)}
        </h1>
      </div>
    );
  }

  renderTag(pkg) {
    return <Tag type="gray">v{pkg.version}</Tag>;
  }

  renderAuthor(pkg) {
    if (isNil(pkg.author) || isNil(pkg.author.name)) {
      return;
    }

    return <div role="author" className={classes.author}>{`By: ${pkg.author.name}`}</div>;
  }
}
