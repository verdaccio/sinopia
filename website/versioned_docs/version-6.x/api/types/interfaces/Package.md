---
id: 'Package'
title: 'Interface: Package'
sidebar_label: 'Package'
sidebar_position: 0
custom_edit_url: null
---

**`Deprecated`**

use Manifest instead

## Properties

### \_attachments

• **\_attachments**: [`AttachMents`](AttachMents.md)

#### Defined in

[packages/core/types/src/manifest.ts:165](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L165)

---

### \_distfiles

• **\_distfiles**: [`DistFiles`](DistFiles.md)

#### Defined in

[packages/core/types/src/manifest.ts:164](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L164)

---

### \_id

• `Optional` **\_id**: `string`

#### Defined in

[packages/core/types/src/manifest.ts:157](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L157)

---

### \_rev

• **\_rev**: `string`

#### Defined in

[packages/core/types/src/manifest.ts:167](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L167)

---

### \_uplinks

• **\_uplinks**: [`UpLinks`](UpLinks.md)

#### Defined in

[packages/core/types/src/manifest.ts:166](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L166)

---

### dist-tags

• **dist-tags**: [`GenericBody`](GenericBody.md)

#### Defined in

[packages/core/types/src/manifest.ts:160](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L160)

---

### name

• **name**: `string`

#### Defined in

[packages/core/types/src/manifest.ts:158](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L158)

---

### readme

• `Optional` **readme**: `string`

#### Defined in

[packages/core/types/src/manifest.ts:162](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L162)

---

### time

• **time**: [`GenericBody`](GenericBody.md)

#### Defined in

[packages/core/types/src/manifest.ts:161](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L161)

---

### users

• `Optional` **users**: [`PackageUsers`](PackageUsers.md)

#### Defined in

[packages/core/types/src/manifest.ts:163](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L163)

---

### versions

• **versions**: [`Versions`](Versions.md)

#### Defined in

[packages/core/types/src/manifest.ts:159](https://github.com/verdaccio/verdaccio/blob/10057a4ff/packages/core/types/src/manifest.ts#L159)
