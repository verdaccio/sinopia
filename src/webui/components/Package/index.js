import React from 'react';
import PropTypes from 'prop-types';
import Chip from '@material-ui/core/Chip';
import { Link } from 'react-router-dom';

import { formatDateDistance } from '../../utils/package';

import classes from './package.scss';

const Package = ({ name, version, author, description, license, time, keywords }) => {
  return (<section className={classes.package}>
    <Link to={`detail/${name}`}>
      <div className={classes.header}>
        <div className={classes.title}>
          <h1>
            {name} <Chip label={`v${version}`} />
          </h1>
        </div>
        <div role="author" className={classes.author}>
          {author ? `By: ${author}` : ''}
        </div>
      </div>
      <div className={classes.footer}>
        <p className={classes.description}>
          {description}
        </p>
      </div>
      <div className={classes.tags}>
        {keywords && keywords.map((keyword, index) => (
          <Chip
            key={index}
            label={keyword}
            className={classes.tag}
          />
        ))}
      </div>
      <div className={classes.details}>
        <div className={classes.homepage}>
          {time ? `Published ${formatDateDistance(time)} ago` : ''}
        </div>
        <div className={classes.license}>
          {license}
        </div>
      </div>
    </Link>
  </section>);
};

Package.propTypes = {
  name: PropTypes.string,
  version: PropTypes.string,
  author: PropTypes.string,
  description: PropTypes.string,
  keywords: PropTypes.array,
  license: PropTypes.string,
  time: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Date)
  ])
};

export default Package;
