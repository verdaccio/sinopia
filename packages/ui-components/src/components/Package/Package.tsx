import { css } from '@emotion/react';
import styled from '@emotion/styled';
import BugReport from '@mui/icons-material/BugReport';
import DownloadIcon from '@mui/icons-material/CloudDownload';
import HomeIcon from '@mui/icons-material/Home';
import { useTheme } from '@mui/material';
import CircularProgress from '@mui/material/CircularProgress';
import Grid from '@mui/material/Grid';
import ListItem from '@mui/material/ListItem';
import Tooltip from '@mui/material/Tooltip';
import React, { useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';

import { Dispatch, Link, RootState, Theme } from '../../';
import { FileBinary, Law, Time, Version } from '../../components/Icons';
import { Author as PackageAuthor, PackageMetaInterface } from '../../types/packageMeta';
import { url, utils } from '../../utils';
import Tag from './Tag';
import {
  Author,
  Avatar,
  Description,
  Details,
  GridRightAligned,
  IconButton,
  OverviewItem,
  PackageListItemText,
  PackageTitle,
  Published,
  TagContainer,
  Wrapper,
  WrapperLink,
} from './styles';

interface Bugs {
  url: string;
}
interface Dist {
  unpackedSize: number;
  tarball: string;
}

export interface PackageInterface {
  name: string;
  version: string;
  time?: number | string;
  author: PackageAuthor;
  description?: string;
  keywords?: string[];
  license?: PackageMetaInterface['latest']['license'];
  homepage?: string;
  bugs?: Bugs;
  dist?: Dist;
  showDownload?: boolean;
}

const Package: React.FC<PackageInterface> = ({
  author: { name: authorName, avatar: authorAvatar },
  bugs,
  description,
  dist,
  homepage,
  keywords = [],
  license,
  name: packageName,
  time,
  showDownload = true,
  version,
}) => {
  const config = useSelector((state: RootState) => state.configuration.config);
  const dispatch = useDispatch<Dispatch>();
  const { t } = useTranslation();
  const theme = useTheme();
  const isLoading = useSelector((state: RootState) => state?.loading?.models.download);

  const handleDownload = useCallback(
    async (tarballDist: string) => {
      // FIXME: check, the dist might be different thant npmjs
      const link = tarballDist.replace(`https://registry.npmjs.org/`, config.base);
      dispatch.download.getTarball({ link });
    },
    [dispatch, config]
  );

  const renderVersionInfo = (): React.ReactNode =>
    version && (
      <OverviewItem>
        <StyledVersion />
        {t('package.version', { version })}
      </OverviewItem>
    );

  const renderAuthorInfo = (): React.ReactNode => {
    const name = utils.getAuthorName(authorName);
    return (
      <Author>
        <Avatar alt={name} src={authorAvatar} />
        <Details>
          <div
            style={{
              fontSize: '12px',
              fontWeight: theme?.fontWeight.semiBold,
              color:
                theme?.palette.mode === 'light' ? theme?.palette.greyLight2 : theme?.palette.white,
            }}
          >
            {name}
          </div>
        </Details>
      </Author>
    );
  };

  const renderFileSize = (): React.ReactNode =>
    dist?.unpackedSize && (
      <OverviewItem>
        <StyledFileBinary />
        {utils.fileSizeSI(dist.unpackedSize)}
      </OverviewItem>
    );

  const renderLicenseInfo = (): React.ReactNode =>
    license && (
      <OverviewItem>
        <StyledLaw />
        {license as string}
      </OverviewItem>
    );

  const renderPublishedInfo = (): React.ReactNode =>
    time && (
      <OverviewItem>
        <StyledTime />
        <Published>{t('package.published-on', { time: utils.formatDate(time) })}</Published>
        {utils.formatDateDistance(time)}
      </OverviewItem>
    );

  const renderHomePageLink = (): React.ReactNode =>
    homepage &&
    url.isURL(homepage) && (
      <Link external={true} to={homepage}>
        <Tooltip aria-label={t('package.homepage')} title={t('package.visit-home-page')}>
          <IconButton aria-label={t('package.homepage')} size="large">
            <HomeIcon />
          </IconButton>
        </Tooltip>
      </Link>
    );

  const renderBugsLink = (): React.ReactNode =>
    bugs?.url &&
    url.isURL(bugs.url) && (
      <Link external={true} to={bugs.url}>
        <Tooltip aria-label={t('package.bugs')} title={t('package.open-an-issue')}>
          <IconButton aria-label={t('package.bugs')} size="large">
            <BugReport />
          </IconButton>
        </Tooltip>
      </Link>
    );

  const renderDownloadLink = (): React.ReactNode =>
    dist?.tarball &&
    url.isURL(dist.tarball) && (
      <Link
        onClick={() => {
          handleDownload(dist.tarball);
        }}
        to="#"
      >
        <Tooltip
          aria-label={t('package.download', { what: t('package.the-tar-file') })}
          title={t('package.tarball')}
        >
          <IconButton aria-label={t('package.download')} size="large">
            {isLoading ? (
              <CircularProgress size={13}>
                <DownloadIcon />
              </CircularProgress>
            ) : (
              <DownloadIcon />
            )}
          </IconButton>
        </Tooltip>
      </Link>
    );

  const renderPrimaryComponent = (): React.ReactNode => {
    return (
      <Grid container={true} item={true} xs={12}>
        <Grid item={true} xs={11}>
          <WrapperLink to={`/-/web/detail/${packageName}`}>
            <PackageTitle className="package-title" data-testid="package-title">
              {packageName}
            </PackageTitle>
          </WrapperLink>
        </Grid>
        <GridRightAligned
          alignItems="center"
          container={true}
          item={true}
          justify="flex-end"
          spacing={3}
          xs={true}
        >
          {renderHomePageLink()}
          {renderBugsLink()}
          {showDownload && renderDownloadLink()}
        </GridRightAligned>
      </Grid>
    );
  };

  const renderSecondaryComponent = (): React.ReactNode => {
    const tags = keywords.sort().map((keyword, index) => <Tag key={index}>{keyword}</Tag>);
    return (
      <>
        <Description>{description}</Description>
        {tags.length > 0 && <TagContainer>{tags}</TagContainer>}
      </>
    );
  };

  const renderPackageListItemText = (): React.ReactNode => (
    <PackageListItemText
      className="package-link"
      primary={renderPrimaryComponent()}
      secondary={renderSecondaryComponent()}
    />
  );

  return (
    <Wrapper className={'package'} data-testid="package-item-list">
      <ListItem alignItems={'flex-start'}>{renderPackageListItemText()}</ListItem>
      <ListItem alignItems={'flex-start'}>
        {renderAuthorInfo()}
        {renderVersionInfo()}
        {renderPublishedInfo()}
        {renderFileSize()}
        {renderLicenseInfo()}
      </ListItem>
    </Wrapper>
  );
};

export default Package;

const iconStyle = ({ theme }: { theme: Theme }) => css`
  margin: 2px 10px 0 0;
  fill: ${theme?.palette.mode === 'light' ? theme?.palette.greyLight2 : theme?.palette.white};
`;

const StyledVersion = styled(Version)`
  ${iconStyle};
`;

const StyledFileBinary = styled(FileBinary)`
  ${iconStyle};
`;

const StyledLaw = styled(Law)`
  ${iconStyle};
`;

const StyledTime = styled(Time)`
  ${iconStyle};
`;
