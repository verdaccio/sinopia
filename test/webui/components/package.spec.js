/**
 * Package component
 */

import React from 'react';
import { mount } from 'enzyme';
import Package from '../../../src/webui/src/components/Package';
import { BrowserRouter } from 'react-router-dom';

describe('<Package /> component', () => {
  it('should load the component', () => {
    const props = {
      name: 'verdaccio',
      version: '1.0.0',
      description: 'Private NPM repository',
      author: { name: 'Sam' }
    };
    const wrapper = mount(
      <BrowserRouter>
        <Package package={props} />
      </BrowserRouter>
    );

    // renderAuthor method
    const renderAuthor = wrapper.find(Package).instance().renderAuthor;
    expect(renderAuthor({ author: {} })).toBeUndefined();
    expect(renderAuthor({ author: { name: 'sam' } })).toBeDefined();

    // integration expectations
    expect(wrapper.find('a').prop('href')).toEqual('detail/verdaccio');
    expect(wrapper.find('h1').text()).toEqual('verdacciov1.0.0');
    expect(wrapper.find('.el-tag--gray').text()).toEqual('v1.0.0');
    expect(
      wrapper
        .find('span')
        .filterWhere(n => n.prop('role') === 'author')
        .text()
    ).toEqual('By: Sam');
    expect(wrapper.find('p').text()).toEqual('Private NPM repository');
  });
});
