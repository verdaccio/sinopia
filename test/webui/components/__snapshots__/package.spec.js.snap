// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Package /> component should load the component 1`] = `"<a class=\\"package\\" href=\\"detail/verdaccio\\"><h1>verdaccio<span class=\\"el-tag el-tag--gray\\">v1.0.0</span></h1><span role=\\"author\\" class=\\"author\\">By: Sam</span><p>Private NPM repository</p></a>"`;
